riffle-server/src/store/mem/buffer.rs

```diff
@@ -247,12 +283,20 @@ impl MemoryBuffer {
     #[trace]
     pub fn append(&self, blocks: Vec<Block>, size: u64) -> Result<()> {
         let mut buffer = self.buffer.lock();
-        let mut staging = &mut buffer.staging;
-        staging.push(blocks);
+        let current_position = buffer.staging.len();
 
+        // Record batch boundary
+        if !blocks.is_empty() {
+            buffer.batch_boundaries.push(current_position);
+        }
+        for (idx, block) in blocks.into_iter().enumerate() {
+            buffer
+                .block_position_index
+                .insert(block.block_id, current_position + idx);
+            buffer.staging.push(block);
```

Comment:
this will cause the frequent vec extend operations, that's expensive in the hot path.